### DISCUSSION TITLE
Tx Power

### DISCUSSION THREAD

**Discussion by facepolish** [hbldh/bleak#984] [Q&A] [answered]
Hi, I want to get Tx Power value to calculate the distance of each devices.

So far I've found a way to get Tx Power information from platform_data. 
   on Mac (Darwin)  ->  advertisement_data.platform_data[1].get("kCBAdvDataTxPowerLevel")
   on Linux  ->  advertisement_data.platform_data.get('TxPower') 
                if the device has Tx Power information. 
However I am not sure how to get Tx Power on Windows. BluetoothLEAdvertisementReceivedEventArgs also has TransmitPowerLevelInDBm as properly. So I think it is possible.  
My question are
 1. Is there a better way (platform independent way ) to get Tx Power information with Bleak?
 2. How can I decode BluetoothLEAdvertisementReceivedEventArgs  class with Bleak with Python?

 Thank you in advance.

**Comment by dlech** [comment, marked as answer]
> Is there a better way (platform independent way ) to get Tx Power information with Bleak?

This would be a good one to add to Bleak so that users don't have to write platform-specific code (i.e. add a `tx_power` attribute to the `bleak.backends.scanner.AdvertisementData` class and populate it as you have described).  Would you be able to make a pull request to implement this?


> How can I decode BluetoothLEAdvertisementReceivedEventArgs class with Bleak with Python?

There is nothing to decode. In the WinRT bindings, the property will be lower snake case (`transmit_power_level_in_d_bm`).

**Reply by facepolish** [reply]
Thank you !!
As for WinRT, I confirmed 
   advertisement_data.platform_data[1].adv.transmit_power_level_in_d_bm
return  Tx Power data.

As for tx_power attribute , I'll try to make changes and make a pull request soon. I appreciate your help very much.

**Comment by MarkusPiotrowski** [comment]
I'm actually writing a small module to allow BLE access on Android via Python when using the Beeware framework. The idea is to have a limited drop in replacement for Bleak, since the Python4Android code from Bleak does not work in Beeware (as discussed elsewhere).

I just stumbled across the Tx Power thing. The code as used by Bleak for the Android platform nearly never returns tx_power, while the code for the Windows platform does. So I got different tx_power results from the same device depending on the platform I use.

I did some reading and must come to the conclusion that `tx_power` is misleading. As far as I have understood from the Bleak code, the Linux, Mac and Window backends all receive **transmission power _level_**, which is not the same as *transmission power*. 
Transmission power _level_ is the set power on the BLE chip (e.g. 12 dBm) and is advertised by some BLE devices.
Transmission power (as returned by the android.bluetooth.le scanner) seems to be the "calibrated transmission power", given as RSSI at a distance of 1 m. This seems to be used by beacons, but nearly never by other BLE devices (at least I didn't find one the last 3 days of scanning around...) BTW, the calibrated transmission power is what is actually required for proper distance calculations.

To be consistent with the other backend APIs, the Android backend should not use `ScanResult.getTxPower()` but `ScanRecord.getTxPowerLevel()` instead, if I'm not mistaken.

And `tx_power_level` probably would have been a better name for the property.

**Reply by dlech** [reply]
Thanks for noticing, I opened a new issue for the bug.

And if it is easier to build and test Python on Android using Beeware than it is with Python4Android, I wouldn't be opposed to having it as an alternative backend here.

**Reply by dlech** [reply]
And if the name is really problematic, I would take a pull request to add a new property and deprecate the old one.

**Comment by MarkusPiotrowski** [comment]
I don't know if the name is really problematic. There are some BLE libraries around that use a method `getTxPower()` to actually retrieve the transmission power _level_. But on Android you have two different methods and I wondered what the difference is (I'm still not 100% what `ScanResult.getTxPower()` in Android should return).

As for another Android backend in Bleak: It is not, that I have not thought about this. But as a hobby programmer, it's, for the moment, much easier for me to write a standalone module than trying to build this into the existing Bleak backend djungel ;-)

**Reply by dlech** [reply]
> (I'm still not 100% what `ScanResult.getTxPower()` in Android should return).

https://stackoverflow.com/a/78280387/1976323 sounds correct to me :wink:
